### ISSUE TITLE
Deduplicate CVEs

### ISSUE THREAD

**Issue by fricklerhandwerk** [Nix-Security-WG/nix-security-tracker#201]
In our ingested data, many CVE numbers appear multiple times since they seem to come from different sources.

As a security team member, I want to only have to deal with unique CVEs.

For example, stage the ingested data before presenting it for triage, and deduplicate it automatically as far as possible:
- Some of the data is redundant but not identical, e.g. differing only in capitalisation; pick one variant consistently
- Some data fields are filled in one variant but not the other; merge
- Some data fields conflict; present for manual resolution

**Comment by fricklerhandwerk** [comment]
Discussed with @RaitoBezarius and @erictapen:
- This issue is blocking UI development for the #203, raised priority
- The main problem is that we don't want duplicate CVE IDs in the suggestions view
- We'll move towards a unidirectional processing pipeline, where merged CVEs are like a cache that can be re-created by a full bulk run
- Merged CVEs should refer to the raw data

**Comment by fricklerhandwerk** [comment]
This seems to be a problem with the suggestion engine.

This condition apparently does not hold:
```python
    if CVEDerivationClusterProposal.objects.filter(cve=container.cve).exists():
        logger.warning(
            "Proposals already exist for '%s', skipping linkage.", container.cve
        )
        return
```

**Comment by RaitoBezarius** [comment]
The root cause is not duplicate proposals but duplicate ingested CVEs.

**Comment by erictapen** [comment]
I just found a weird thing which I think is related.

I *am* seeing duplicate proposals in the suggestions list.

![tmp 0vtBs2tcYd](https://github.com/user-attachments/assets/dcb2c20e-eb26-496e-90df-d753b5f07559)

Note how the `suggestion_id` is `864` in both cases. This directly represents `pk` field in the ListView, showing `CVEDerivationClusterProposal`s. I wonder how this is even possible, that we get duplicated entries with the same primary key? Which curiously even show different affected package names?

The code running is my branch `remove-package-selection`, though I doubt it's specific to my changes.

**Comment by RaitoBezarius** [comment]
It can occur due to a JOIN that repeats some fields.

**Comment by fricklerhandwerk** [comment]
`queryset2` is what the view is currently doing:

```
In [28]: queryset = CVEDerivationClusterProposal.objects.select_related("cve").filter(status=CVEDerivationClusterProposal.Status.PENDING).prefetch_related("
    ...: derivations","derivations__parent_evaluation")

In [29]: queryset2 = queryset.annotate(
    ...:     package_name=F("cve__container__affected__package_name"),
    ...:     base_severity=Coalesce(
    ...:         F("cve__container__metrics__base_severity"), Value(Severity.NONE)
    ...:     ),
    ...:     title=F("cve__container__title"),
    ...:     description=F("cve__container__descriptions__value"),
    ...: )

In [30]: queryset.count()
Out[30]: 8

In [31]: queryset2.count()
Out[31]: 66881
```

**Comment by fricklerhandwerk** [comment]
It seems the simplest way to both solve this issue and keep the query time down despite a suggestion containing lots of information is caching the entire suggestion (including all related models, filtered down to what we need) by ID, already at creation time. Suggestions only change once we touch them, and even then we could cache everything except the moving parts (currently the list of linked derivations).

Then the view query can only get the item IDs, and the rest is taken directly from the cache. Opened #377 

@alejandrosame btw. we may want to do the same for activity logs: #366

**Comment by fricklerhandwerk** [comment]
Actually it doesn't really matter *how* we cache those aggregate data, it could just as well be in a helper table.

We discussed 2-3 weeks ago with @RaitoBezarius and @erictapen that we'd probably want the data to flow across intermediate tables in a way that can be recomputed if needed, but accessed quickly when it's available.

**Comment by fricklerhandwerk** [comment]
I think this is resolved now with #378, but I need to try it out.

**Comment by fricklerhandwerk** [comment]
This seems fixed.
